lib/run_part.c

```diff
@@ -0,0 +1,86 @@
+#include <dirent.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+int run_part (char *script_path, char *name, char *system)
+{
+	int pid;
+	int wait_status;
+	int pid_status;
+	char *args[] = { script_path, NULL };
+
+	pid=fork();
+	if (pid==-1){
+		perror ("Could not fork");
+		return 1;
+	}
+	if (pid==0) {
+		setenv (system,name,1);
+		execv (script_path,args);
+		perror ("execv");
+		exit(1);
+	}
+
+	pid_status = wait (&wait_status);
+	if (pid_status == pid) {
+		return (wait_status);
+	}
+
+	perror ("waitpid");
+	return (1);
+}
+
+int run_parts (char *directory, char *name, char *system)
+{
+	struct dirent **namelist;
+	int scanlist;
+	int n;
+	int execute_result;
+
+	scanlist = scandir (directory, &namelist, 0, alphasort);
+	if (scanlist<0) {
+		return (0);
+	}
+
+	for (n=0; n<scanlist; n++) {
+		int path_length;
+		struct stat sb;
+
+		path_length=strlen(directory) + strlen(namelist[n]->d_name) + 2;
+		char *s = (char*)malloc(path_length);
+		if (!s) {
+			printf ("could not allocate memory\n");
+			return (1);
+		}
+		snprintf (s, path_length, "%s/%s", directory, namelist[n]->d_name);
+
+		execute_result = 0;
+		if( stat (s, &sb) == -1 ) {
+			free (s);
+			free (namelist[n]);
+			perror ("stat");
+			return (1);
+		}
+
+		if (S_ISREG (sb.st_mode)) {
+			execute_result = run_part (s, name, system);
+		}
+
+		free (s);
+		free (namelist[n]);
+
+		if (execute_result!=0) {
+			break;
```

Comment:
In this case you'll leak namelist[n+1..scanlist]

Reply:
Good point, this should be fixed now.